autoload/go/rename.vim

```diff
@@ -41,63 +41,34 @@ function! go#rename#Rename(bang, ...) abort
 endfunction
 
 function s:rename_job(args)
-  let state = {
-        \ 'exited': 0,
-        \ 'closed': 0,
-        \ 'exitval': 0,
-        \ 'messages': [],
-        \ 'status_dir': expand('%:p:h'),
-        \ 'bang': a:args.bang
-      \ }
-
-  function! s:callback(chan, msg) dict
-    call add(self.messages, a:msg)
-  endfunction
-
-  function! s:exit_cb(job, exitval) dict
-    let self.exited = 1
-    let self.exitval = a:exitval
-
-    let status = {
-          \ 'desc': 'last status',
-          \ 'type': "gorename",
-          \ 'state': "finished",
-          \ }
-
-    if a:exitval
-      let status.state = "failed"
-    endif
-
-    call go#statusline#Update(self.status_dir, status)
+  let l:job_opts = {
+        \ 'bang': a:args.bang,
+        \ 'for': 'GoRename',
+        \ 'statustype': 'gorename',
+        \ }
 
-    if self.closed
-      call s:parse_errors(self.exitval, self.bang, self.messages)
-    endif
-  endfunction
+  " autowrite is not enabled for jobs
+  call go#cmd#autowrite()
+  let l:cbs = go#job#Options(l:job_opts)
 
-  function! s:close_cb(ch) dict
-    let self.closed = 1
+  " wrap l:cbs.exit_cb in s:exit_cb.
+  let l:cbs.exit_cb = funcref('s:exit_cb', [l:cbs.exit_cb])
 
-    if self.exited
-      call s:parse_errors(self.exitval, self.bang, self.messages)
-    endif
-  endfunction
-
-  " explicitly bind the callbacks to state so that self within them always
-  " refers to state. See :help Partial for more information.
-  let start_options = {
-        \ 'callback': funcref("s:callback", [], state),
-        \ 'exit_cb': funcref("s:exit_cb", [], state),
-        \ 'close_cb': funcref("s:close_cb", [], state),
-        \ }
+  call go#job#Start(a:args.cmd, l:cbs)
+endfunction
 
-  call go#statusline#Update(state.status_dir, {
-        \ 'desc': "current status",
-        \ 'type': "gorename",
-        \ 'state': "started",
-        \})
+" s:exit_cb reloads any changed buffers and then calls next.
+function! s:exit_cb(next, job, exitval)
+  " reload all files to reflect the new changes. We explicitly call
+  " checktime to trigger a reload of all files. See
+  " http://www.mail-archive.com/vim@vim.org/msg05900.html for more info
+  " about the autoread bug
+  let current_autoread = &autoread
+  set autoread
+  silent! checktime
+  let &autoread = current_autoread
 
-  call job_start(a:args.cmd, start_options)
+  call call(a:next, [a:job, a:exitval])
```

Comment:
Why do we have here two call's? I think I saw this once but I'm not sure to be honest what it means in this context.

Reply:
The first `call` is  required to call the function named `call`. We have to use the function named `call` to execute the funcref named `a:next`.
 
See :help :call` and `help call()`.